### PR TITLE
avoid holding the toppar lock across a write syscall

### PR DESCRIPTION
**Pull request by tbsaunde**
Holding locks across syscalls is generally not a great idea since it
means spending longer with the lock and increases the chances the
process will spend time holding the lock but descheduled.  In this case
it is especially unfortunate because the write will make a broker thread
runnable if it is blocked on poll, but if the broker thread runs before
the thread calling write() it will immediately block again trying to
take the toppar lock.  This appears to reduce worst case latency by at
least 1ms, and quiet possibly much more, its hard to know exactly how
much should be attributed to this instead of other freek events hurting
worst case latency.  Its somewhat unfortunate we need to do a little
more work unconditionally, but the atomic accesses can be optimized
somewhat, and it may be a good idea to use the wakeup fd system more.

### CODE REVIEW

File: src/rdkafka_partition.c [hunk 1]
```diff
@@ -609,26 +609,23 @@ void rd_kafka_toppar_desired_del (rd_kafka_toppar_t *rktp) {
  * Append message at tail of 'rktp' message queue.
  */
 void rd_kafka_toppar_enq_msg (rd_kafka_toppar_t *rktp, rd_kafka_msg_t *rkm) {
+	int wakeup_fd, queue_len;
+	rd_kafka_t *rk;
```
Comment (edenhill) [inline src/rdkafka_partition.c]: You dont need to keep `rk`, use rd_assert() which does not take an rk argument.

File: src/rdkafka_partition.c [hunk 2]
```diff
@@ -609,26 +609,23 @@ void rd_kafka_toppar_desired_del (rd_kafka_toppar_t *rktp) {
...
+        rd_kafka_toppar_unlock(rktp);
 #ifndef _MSC_VER
-        if (rktp->rktp_msgq_wakeup_fd != -1 &&
-            rd_kafka_msgq_len(&rktp->rktp_msgq) == 1) {
+        if (wakeup_fd >= 0 && queue_len == 1) {
                 char one = 1;
                 int r;
-                r = rd_write(rktp->rktp_msgq_wakeup_fd, &one, sizeof(one));
-                if (r == -1)
-                        rd_kafka_log(rktp->rktp_rkt->rkt_rk, LOG_ERR, "PARTENQ",
```
Comment (edenhill) [inline src/rdkafka_partition.c]: why was this removed?

File: src/rdkafka_partition.c [hunk 3]
```diff
@@ -609,26 +609,23 @@ void rd_kafka_toppar_desired_del (rd_kafka_toppar_t *rktp) {
...
 	rd_kafka_toppar_lock(rktp);
+	wakeup_fd = rktp->rktp_msgq_wakeup_fd;
+	rk = rktp->rktp_rkt->rkt_rk;
 	rd_kafka_msgq_enq(&rktp->rktp_msgq, rkm);
+	queue_len = rd_kafka_msgq_len(&rktp->rktp_msgq);
+        rd_kafka_toppar_unlock(rktp);
 #ifndef _MSC_VER
-        if (rktp->rktp_msgq_wakeup_fd != -1 &&
-            rd_kafka_msgq_len(&rktp->rktp_msgq) == 1) {
+        if (wakeup_fd >= 0 && queue_len == 1) {
```
Comment (edenhill) [inline src/rdkafka_partition.c]: This is not completely without side-effects:

Having the queue len + write outside of holding the lock risks two (or more) threads calling write() simultaneously which will cause additional wake-ups in the broker thread and more context switches in these enq threads, all resulting in increased CPU usage.

Comment (edenhill) [inline src/rdkafka_partition.c]: Please try to keep as much of the existing code as possible if a change is not strictly needed.
E.g.. `wakeup_fd == -1 ..`

I think we should change msgq_len() to return the return-value from atomic_add() so that we don't need the msgq_len() call at all.

Comment (tbsaunde) [reply src/rdkafka_partition.c]: the change to the comparison was an accident, and I'll revert that, though I do sort of think >= 0 is a better test given that other negative return values don't mean anything.
I was planning on doing that change to msgq_enq, but wanted to keep it separate, and didn't think of it till this was written.

File: src/rdkafka_partition.c (hunk 2 above)
Comment (tbsaunde) [reply src/rdkafka_partition.c]: I replaced it with the assert because I can't see an error that can actually happen except EINTER maybe which maybe we should separately handle.  I don't have a problem with the logging in general, but in this case it makes it hard to drop the lock earlier.  Since the log message includes the topic name we somehow need to have access to that in the exceptional case.  If we drop the lock before write() we don't know that string will even still be around right? so the alternative I can see is copying the string every time while holding the lock and that seems really bad.  I guess we could drop the topic name / partition from the message, but that would mean still fetching rk while holding the lock.

File: src/rdkafka_partition.c (hunk 3 above)
Comment (tbsaunde) [reply src/rdkafka_partition.c]: I don't think so?  The key thing is that the queue length is read while we hold the lock.  That means if two threads read a queue length of 1 in the time between when those two threads held the lock another thread must have dequeued messages to set the queue length to 0.  Another way to think about this is that if you put the unlock after the if the additional operations that take place with the lock held  only depend on values local to the executing thread.

File: src/rdkafka_partition.c (hunk 1 above)
Comment (tbsaunde) [reply src/rdkafka_partition.c]: ah thanks, though this may conflict with your comments on the logging change below?

**Comment by tbsaunde** [comment]
@edenhill, I believe I've addressed all your comments, can you please take another look?

File: src/rdkafka_partition.c (hunk 2 above)
Comment (edenhill) [reply src/rdkafka_partition.c]: We can't make any safe assumptions on what error we get from a system call (since the code runs on a plethora of different operating systems and versions), so we must be very careful in not doing anything drastic in these rare unknown cases.
Thus, we absolute don't want to crash the application if a non-essential wakeup file descriptor is not writable :)

Generally, asserts should only be used to guard for internal state and should never be used for external input.

Regarding the log locking:
The log message is safe without locking, because:
 * since all relevant objects (rktp, rkt, rk) are refcounted and our caller must hold a reference these objects will not go away from under our feet
 * `rktp->rktp_rkt` and `rkt->rkt_rk` are constant, they are never changed during the lifetime of these objects
 * same with `rkt_topic`, `rktp_partition`, etc.

The only thing that may change is `rktp->rktp_msgq_wakeup_fd` which must be protected by a lock, but since you've already extracted the fd into a variable we can use that instead.
This also shows us a valid scenario where the write may fail:
 * thread 1 is in enq_msg() having extracted the fd and getting ready to write
 * thread 2 removes the io_event fd from the queue and closes the fd
In this case thread 1's write() will fail, but the program should continue to operate since on the next call the enq_msg() the fd will no longer be set.

File: src/rdkafka_topic.c [hunk 4]
```diff
@@ -657,7 +657,7 @@ static void rd_kafka_topic_assign_uas (rd_kafka_itopic_t *rkt,
 	rd_kafka_msg_t *rkm, *tmp;
 	rd_kafka_msgq_t uas = RD_KAFKA_MSGQ_INITIALIZER(uas);
 	rd_kafka_msgq_t failed = RD_KAFKA_MSGQ_INITIALIZER(failed);
-	int cnt;
+	int cnt, failed_cnt = 0;
```
Comment (edenhill) [inline src/rdkafka_topic.c]: This is the slow-path so you don't really need to do this optimization here. Suggest removing this patch.

File: src/rdkafka_msg.c [hunk 5]
```diff
@@ -482,17 +482,18 @@ int rd_kafka_msgq_age_scan (rd_kafka_msgq_t *rkmq,
...
 	TAILQ_FOREACH_SAFE(rkm, &rkmq->rkmq_msgs, rkm_link, tmp) {
 		if (likely(rkm->rkm_ts_timeout > now))
 			break;
 
 		rd_kafka_msgq_deq(rkmq, rkm, 1);
-		rd_kafka_msgq_enq(timedout, rkm);
+		timedout_cnt = rd_kafka_msgq_enq(timedout, rkm);
 	}
 
-	return rd_atomic32_get(&timedout->rkmq_msg_cnt) - cnt;
```
[the remark this answers is not in the document]
Comment (tbsaunde) [reply src/rdkafka_msg.c]: Its O(N), but one would hope in practice nothing times out and its ~O(1).  aiui atomic operations are especially costly, so if we generally don't loop it seems like it may actually matter.  I'm pretty sure the calculation is already racey and shouldn't be used for something that must know exactly how many elements are in the queue at this exact moment, I think to know exactly you need to hold the queue lock.  The problem is that between when the atomic fetch is performed and the value is returned to the caller another thread could modify timedout.  Looking at the rdkafka_broker.c caller it seems ok with this raceyness, in the worst case you miss the new queue elements, but get them the next time through.  It looks like the only other caller only uses the return value in logging, so that's not terribly important.

That all said I'm happy to revert it doing the conservatively correct thing until we have numbers showing it matters.  Mostly this second patch was because I think you suggested it earlier.

File: src/rdkafka_topic.c (hunk 4 above)
Comment (tbsaunde) [reply src/rdkafka_topic.c]: Well, the if (failed_cnt > 0) check at the very end is always executed, so that matters somewhat I guess.  It also moves us closer to a world where the count is just protected by the queue lock instead of using atomic operations.  I'm not sure what downside you see, but I'll remove this files changes for now to get the rest in.

**Comment by edenhill** [comment]
Thanks for this!
